### DISCUSSION TITLE
Why are both `.js` and `.html` files being output for `getStaticProps`?

### DISCUSSION THREAD

**Discussion by geelen** [vercel/next.js#15631] [Help]
So... weird behaviour I've encountered with NextJS's new static-first focus. For a project with the following structure:

```
> tree pages
pages
├── _app.js
├── _document.js
├── _error.js
├── about
│   └── index.js
├── blog
│   ├── [slug].js
│   └── index.js
...
```

I get this `.next/serverless/pages` structure.

```
> tree .next/serverless/pages
.next/serverless/pages
├── _error.js
├── about.html
├── blog
│   ├── [slug].js
│   ├── why-my-cat-is-best.html
│   ├── why-my-cat-is-best.json
│   ├── why-dogs-are-overrated.html
│   └── why-dogs-are-overrated.json
├── blog.html
├── blog.js     # <---
├── blog.json
...
```

Now, the way `blog/[slug].js` vs the prerendered posts make sense, since that's using `getStaticPaths` which I assume can be dynamic, but I'm a bit suprised to see `blog.js` there.

For reference, this is the `getStaticProps` logic I have in `blog/index.js`...

```js
// blog/index.js
import { request } from "../../lib/datocms"

export default function BlogIndex(props) {
  const posts = this.props.posts.allBlogPosts

  return (
    //..
  )
}

export async function getStaticProps() {
  const blogPostList = `
    query blogPostList {
      allBlogPosts {
        title
        slug
      }
    }
  `

  const posts = await request({
    query: blogPostList
  })

  return {
    props: { posts }
  }
}
```

From my understanding, `getStaticProps` run only at build time, so this should mean that only `.next/serverless/pages/blog.html` is generated, not `.next/serverless/pages/blog.js`, yet both are there. Does anyone know why both are outputted?

Does every file that uses `getStaticProps` generate a `.js` file on the serverless build? And is that designed to allow the hosting platform to refresh the static snapshot? I thought you had to opt-in to that behaviour...

**Comment by ijjk** [comment]
The server JS file is always output for pages with `getStaticProps` since it is needed to prerender the page during the build, it is also used when leveraging [preview mode](https://nextjs.org/docs/advanced-features/preview-mode). If you don't leverage API routes (required for enabling preview mode) and the pages are completely static (don't leverage fallback or revalidate support) with `getStaticProps`, lambdas won't be deployed for these pages on Vercel anymore though
